### PR TITLE
#439 Carbone template for Feature Showcase

### LINKED ISSUE #439: Generate a Carbone document template for an application with all responses by applicant and review_responses

### PR DESCRIPTION
**Pull request by CarlosNZ**
Fixes #439 

Having done this, I would say it's definitely too complicated to try and make a "general" carbone template for all templates. Every question is different -- not just different question types, but different instances of the same question type require case-specific formatting.

Here is the carbone template for Feature Showcase: [feature_showcase.odt](https://github.com/openmsupply/application-manager-server/files/6875752/feature_showcase.odt)

Here is an example output PDF: [feature_showcase_S-VKT-0100_UlUggoa7IIjW0cz9urfrN.pdf](https://github.com/openmsupply/application-manager-server/files/6875809/feature_showcase_S-VKT-0100_UlUggoa7IIjW0cz9urfrN.pdf)

### How to test

Easiest way is to use Postman to manually run the Actions. Here is a collection with the endpoints and parameters you'll need: 
[Application Manager.postman_collection.json.zip](https://github.com/openmsupply/application-manager-server/files/6875759/Application.Manager.postman_collection.json.zip)

1. Start a new Feature Showcase application. You don't need to submit it, just make sure responses have been added (it's probably best not to submit, so you can quickly change responses and see how the PDF output changes)
2. Use the `/upload` endpoint to upload the carbone template file and return a uniqueId
3. Use the `/run-action` endpoint  to "generateDoc", using your application ID and the previous uniqueId (for the file), like so:
```js
{
    "actionCode": "generateDoc",
    "applicationId": 1,
    "parameters": { 
        "docTemplateId": "vXQzU3H3EcFOOSmgiDady"
    }
}
```
4. Open the created PDF and observe the results :) 

### Additional changes

- added an "additionalData" parameter to "generateDoc" (similar to "generateTextString"), in case we want to pass data other than "applicationData" and "outputCumulative"
- "applicationData" now contains "orgName" as well (acquired as part of `getUserData`)
- fix for "subfolder" parameter in `/upload` endpoint (can use either "subfolder" or "subFolder", or "sub_folder")

### CODE REVIEW

**Comment by nmadruga** [comment]
> 3\. Use the `/run-action` endpoint  to "generateDoc", using your application ID and the previous uniqueId (for the file), like so:
> 
> ```js
> {
>     "actionCode": "generateDoc",
>     "applicationId": 1,
>     "parameters": { 
>         "docTemplateId": "vXQzU3H3EcFOOSmgiDady"
>     }
> }
> ```

I'm getting an error when trying to run `generateDoc` action : (
Screen Shot 2021-08-06 at 10 18 29 AM

Setup:
1)
Screen Shot 2021-08-06 at 10 18 50 AM
2)
Screen Shot 2021-08-06 at 10 18 39 AM

**Comment by nmadruga** [comment]
All good - problem was related to format used in Postman (in a older version probably didn't import correctly):
 
Screen Shot 2021-08-06 at 10 18 39 AM

**Review by nmadruga — APPROVED** [review]
Thanks this look great. Also added the issue to update the Feature showcase (with slight changes in codes and moved elements) in our new snapshot_basic #497 later

File: src/components/files/fileHandler.ts [hunk 1]
```diff
@@ -52,8 +52,9 @@ export async function saveFiles(data: any, queryParams: HttpQueryParameters) {
       const ext = path.extname(file.filename)
       const basename = path.basename(file.filename, ext)
       const unique_id = queryParams?.uid ?? nanoid()
-
-      const subfolder = queryParams?.subFolder ?? queryParams?.application_serial ?? ''
+      console.log('queryParams', queryParams)
```
[the remark this answers is not in the document]
Comment (CarlosNZ) [reply src/components/files/fileHandler.ts]: 0d0c0457

File: src/components/postgresConnect.ts [hunk 2]
```diff
@@ -433,9 +433,18 @@ class PostgresDB {
       FROM "user"
       WHERE id = $1
       `
+    const text2 = `
+      SELECT name as "orgName"
+      FROM organisation
+      WHERE id = $1
+      `
```
[the remark this answers is not in the document]
Comment (CarlosNZ) [reply src/components/postgresConnect.ts:440]: Not sure what you mean.

**Comment by CarlosNZ** [comment]
@nmadruga I'll merge this now for the core functionality, but I'll need to co-ordinate with you regarding how to migrate the .odt file over to your snapshot branch, and making sure we keep the file as part of the FeatureShowcase template.

**Comment by nmadruga** [comment]
> @nmadruga I'll merge this now for the core functionality, but I'll need to co-ordinate with you regarding how to migrate the .odt file over to your snapshot branch, and making sure we keep the file as part of the FeatureShowcase template.

Yes, no problem. Issue for doing this was created already #497

**Pull request merged**